File: android/framework_old/src/main/java/com/dreampany/framework/util/NumberUtil.java
/*
package com.dreampany.framework.util;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.text.DecimalFormat;
import java.util.Locale;
import java.util.Map;
import java.util.NavigableMap;
import java.util.Random;
import java.util.TreeMap;

*/
/**
 * Created by Hawladar Roman on 7/30/2018.
 * BJIT Group
 * hawladar.roman@bjitgroup.com
 *//*

public final class NumberUtil {
    private NumberUtil() {
    }

    private static final NavigableMap<Long, String> suffixes = new TreeMap<>();
    private static Random random = new Random();

    static {
        suffixes.put(1_000L, "K");
        suffixes.put(1_000_000L, "M");
        suffixes.put(1_000_000_000L, "G");
        suffixes.put(1_000_000_000_000L, "T");
        suffixes.put(1_000_000_000_000_000L, "P");
        suffixes.put(1_000_000_000_000_000_000L, "E");
    }

    public static String toNumInUnits(long bytes) {
        int u = 0;
        for (; bytes > 1024 * 1024; bytes >>= 10) {
            u++;
        }
        if (bytes > 1024)
            u++;
        return String.format(Locale.ENGLISH, "%.1f %cB", bytes / 1024f, " kMGTPE".charAt(u));
    }

    public static String readableFileSize(long size) {
        if (size <= 0) return "0";
        final String[] units = new String[]{"B", "kB", "MB", "GB", "TB"};
        int digitGroups = (int) (Math.log10(size) / Math.log10(1024));
        return new DecimalFormat("#,##0.#").format(size / Math.pow(1024, digitGroups)) + " " + units[digitGroups];
    }

    public static String formatCount(long count) {
        //Long.MIN_VALUE == -Long.MIN_VALUE so we need an adjustment here
        if (count == Long.MIN_VALUE) return formatCount(Long.MIN_VALUE + 1);
        if (count < 0) return "-" + formatCount(-count);
        if (count < 1000) return Long.toString(count); //deal with easy case

        Map.Entry<Long, String> e = suffixes.floorEntry(count);
        Long divideBy = e.getKey();
        String suffix = e.getValue();

        long truncated = count / (divideBy / 10); //the number part of the output times 10
        boolean hasDecimal = truncated < 100 && (truncated / 10d) != (truncated / 10);
        return hasDecimal ? (truncated / 10d) + suffix : (truncated / 10) + suffix;
    }

    public static String format(long number) {
        //  Locale.setDefault(new Locale("pl", "PL"));
        String formattedNumber = String.format(Locale.getDefault(), "%,d", number);
        return formattedNumber;
    }

    public static int nextRand(int upper) {
        if (upper <= 0) return -1;
        return random.nextInt(upper);
    }

    public static long nextRand(long upper) {
        if (upper <= 0) return -1;
        return random.nextInt((int)upper);
    }

    public static int nextRand(int min, int max) {
        return random.nextInt(max - min + 1) + min;
    }

    public static int getPercentage(long current, long total) {
        float ratio = (float) current / total;

        int percentage = (int) (ratio * 100);

        if (percentage > 100) {
            percentage = 100;
        }

        return percentage;
    }

    public static double round(double value, int places) {
        if (places < 0) throw new IllegalArgumentException();

        BigDecimal bd = new BigDecimal(value);
        bd = bd.setScale(places, RoundingMode.HALF_UP);
        return bd.doubleValue();
    }

    public static int percentage(int current, int total) {
        int percent = (current * 100) / total;
        return percent;
    }
}
*/

File: android/framework_old/src/main/kotlin/com/dreampany/framework/util/NumberUtil.kt
package com.dreampany.framework.util

import com.google.i18n.phonenumbers.NumberParseException
import com.google.i18n.phonenumbers.PhoneNumberUtil
import timber.log.Timber
import java.math.BigDecimal
import java.math.RoundingMode
import java.text.DecimalFormat
import java.util.*


/**
 * Created by roman on 2019-10-14
 * Copyright (c) 2019 bjit. All rights reserved.
 * hawladar.roman@bjitgroup.com
 * Last modified $file.lastModified
 */
class NumberUtil {
    companion object {
        private val suffixes = TreeMap<Long, String>()
        private val random = Random()

        init {
            suffixes[1_000L] = "K"
            suffixes[1_000_000L] = "M"
            suffixes[1_000_000_000L] = "G"
            suffixes[1_000_000_000_000L] = "T"
            suffixes[1_000_000_000_000_000L] = "P"
            suffixes[1_000_000_000_000_000_000L] = "E"
        }

        fun toNumInUnits(bytes: Long): String {
            var bytes = bytes
            var u = 0
            while (bytes > 1024 * 1024) {
                u++
                bytes = bytes shr 10
            }
            if (bytes > 1024)
                u++
            return String.format(Locale.ENGLISH, "%.1f %cB", bytes / 1024f, " kMGTPE"[u])
        }

        fun readableFileSize(size: Long): String {
            if (size <= 0) return "0"
            val units = arrayOf("B", "kB", "MB", "GB", "TB")
            val digitGroups = (Math.log10(size.toDouble()) / Math.log10(1024.0)).toInt()
            return DecimalFormat("#,##0.#").format(
                size / Math.pow(
                    1024.0,
                    digitGroups.toDouble()
                )
            ) + " " + units[digitGroups]
        }

        fun formatCount(count: Long): String {
            //Long.MIN_VALUE == -Long.MIN_VALUE so we need an adjustment here
            if (count == java.lang.Long.MIN_VALUE) return formatCount(java.lang.Long.MIN_VALUE + 1)
            if (count < 0) return "-" + formatCount(-count)
            if (count < 1000) return java.lang.Long.toString(count) //deal with easy case

            val e = suffixes.floorEntry(count)
            val divideBy = e!!.key
            val suffix = e.value

            val truncated = count / (divideBy!! / 10) //the number part of the output times 10
            val hasDecimal = truncated < 100 && truncated / 10.0 != (truncated / 10).toDouble()
            return if (hasDecimal) (truncated / 10.0).toString() + suffix else (truncated / 10).toString() + suffix
        }

        fun format(number: Long): String {
            //  Locale.setDefault(new Locale("pl", "PL"));
            return String.format(Locale.getDefault(), "%,d", number)
        }

        fun formatSpeed(speed: Int, formatStr: String): String {
            var speedValue = speed.toDouble() / 1048576
            speedValue = BigDecimal(speedValue).setScale(3, RoundingMode.UP).toDouble()
            return String.format(formatStr, speedValue)
        }

        fun nextRand(upper: Int): Int {
            return if (upper <= 0) -1 else random.nextInt(upper)
        }

        fun nextRand(upper: Long): Long {
            return if (upper <= 0) -1 else random.nextInt(upper.toInt()).toLong()
        }

        fun nextRand(min: Int, max: Int): Int {
            return random.nextInt(max - min + 1) + min
        }

        fun getPercentage(current: Long, total: Long): Int {
            val ratio = current.toFloat() / total

            var percentage = (ratio * 100).toInt()

            if (percentage > 100) {
                percentage = 100
            }

            return percentage
        }

        fun round(value: Double, places: Int): Double {
            require(places >= 0)

            var bd = BigDecimal(value)
            bd = bd.setScale(places, RoundingMode.HALF_UP)
            return bd.toDouble()
        }

        fun percentage(current: Int, total: Int): Int {
            return current * 100 / total
        }

        fun parseInt(value: Any?, defaultValue: Int): Int {
            return parseInt(value?.toString(), defaultValue)
        }

        fun parseInt(value: String?, defaultValue: Int): Int {
            try {
                return value?.toInt() ?: defaultValue
            } catch (e: NumberFormatException) {
                return defaultValue
            }
        }

        fun randomBool(): Boolean {
            return random.nextBoolean()
        }

        fun isValidPhoneNumber(countryCode: String, number: String): Boolean {
            val phoneUtil = PhoneNumberUtil.getInstance()
            try {
                val phone = phoneUtil.parse(number, countryCode)
                return phoneUtil.isValidNumber(phone)
            } catch (error: NumberParseException) {
                Timber.e(error)
            }
            return false
        }
    }
}